[Pytorch] Consolidate Strobelight compile time profiler between OSS and fbcode

kollasb opened the pull request:
Summary:
Move towards consolidating strobelight profiler implementations between OSS and fbcode. This change is a first step towards that.

- Created a new function to abstract out compile time profiling enablement. This function allows profiler to switch between different function profilers (e.g. Thrift based or CLI based)
- Both OSS and Fbcode now use one compile time profiler in torch/_strobelight

Test Plan:
Tested OSS with following commands:
```
python torch/_strobelight/examples/compile_time_profile_example.py
python torch/_strobelight/examples/cli_function_profiler_example.py

TORCH_COMPILE_STROBELIGHT=TRUE TORCHINDUCTOR_FORCE_DISABLE_CACHES=1 python benchmarks/dynamo/huggingface.py --ci --accuracy --timing --explain --inductor --device cuda --training --amp  --only XLNetLMHeadModel
```

See test commands for fbcode in comments.

Differential Revision: D62444551

laithsakka commented on torch/_strobelight/examples/compile_time_profile_example.py:
```diff
@@ -1,11 +1,10 @@
 
 
 if __name__ == "__main__":
-    # You can pass TORCH_COMPILE_STROBELIGHT=True instead.
```
I guess StrobelightCompileTimeProfiler.enable() will work also that way it ended up being.

kollasb replied:
Yes, that is correct. 

Reason to move to enable_* here to be consistent so that any future changes on enablement can be all handled inside the enable_ function.

kollasb committed:
[Pytorch] Consolidate Strobelight compile time profiler between OSS and fbcode (#135953)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/135953

Move towards consolidating strobelight profiler implementations between OSS and fbcode. This change is a first step towards that.

- Created a new function to abstract out compile time profiling enablement. This function allows profiler to switch between different function profilers (e.g. Thrift based or CLI based)
- Both OSS and Fbcode now use one compile time profiler in torch/_strobelight

Test Plan:
Tested OSS with following commands:
```
python torch/_strobelight/examples/compile_time_profile_example.py
python torch/_strobelight/examples/cli_function_profiler_example.py

TORCH_COMPILE_STROBELIGHT=TRUE TORCHINDUCTOR_FORCE_DISABLE_CACHES=1 python benchmarks/dynamo/huggingface.py --ci --accuracy --timing --explain --inductor --device cuda --training --amp  --only XLNetLMHeadModel
```

See test commands for fbcode in comments.

Reviewed By: laithsakka

Differential Revision: D62444551

kollasb commented:
@pytorchbot revert -m "Broke internal module factory compatibility, revert from Phabricator failed"  -c nosignal

kollasb commented:
For context, following automatic revert from internal PR failed, hence the manual revert.

https://github.com/pytorch/pytorch/pull/136108
